dlib/clustering/bottom_up_cluster.h

```diff
@@ -44,24 +44,47 @@ namespace dlib
 
 // ----------------------------------------------------------------------------------------
 
+    struct bu_cluster
+    {
+        long id;
+        long left;
+        long right;
+        std::vector<long> orig;
+    };
+
+    inline std::ostream& operator<< (std::ostream& out, const bu_cluster& c)
+    {
+        out << "["<<c.id<<","<<c.left<<","<<c.right<<",[";
+        for (auto iter = c.orig.begin(); iter != c.orig.end(); iter++) {
+          out << *iter;
+          if (iter != c.orig.end() - 1)
+            out << ", ";
+        }
+        out <<"]]";
+        return out;
+    }
+
+    static std::vector<bu_cluster> history_;
```

Comment:
You can't stuff a static variable like this into the header file.  I mean you can. But you shouldn't.